app/scripts/lib/tx-verification/tx-verification-middleware.ts

```diff
@@ -0,0 +1,111 @@
+import { hashMessage } from '@ethersproject/hash';
+import { verifyMessage } from '@ethersproject/wallet';
+import type { NetworkController } from '@metamask/network-controller';
+import { rpcErrors } from '@metamask/rpc-errors';
+import { Json, JsonRpcParams, hasProperty, isObject } from '@metamask/utils';
+import {
+  JsonRpcRequest,
+  JsonRpcResponse,
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from 'json-rpc-engine';
+import {
+  SIG_LEN,
+  TRUSTED_BRIDGE_SIGNER,
+} from '../../../../shared/constants/bridge';
+import { FIRST_PARTY_CONTRACT_NAMES } from '../../../../shared/constants/first-party-contracts';
+
+export type BridgeTxParams = {
+  chainId?: `0x${string}`;
+  data: string;
+  from: string;
+  to: string;
+  value: string;
+};
+
+/**
+ * Creates a middleware function that verifies bridge transactions from the
+ * Portfolio.
+ *
+ * @param networkController - The network controller instance.
+ * @returns The middleware function.
+ */
+export function createTxVerificationMiddleware(
+  networkController: NetworkController,
+) {
+  return function txVerificationMiddleware(
+    req: JsonRpcRequest<JsonRpcParams>,
+    _res: JsonRpcResponse<Json>,
+    next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+  ) {
+    if (
+      req.method !== 'eth_sendTransaction' ||
+      !Array.isArray(req.params) ||
+      !isValidParams(req.params)
+    ) {
+      return next();
+    }
+
+    // the tx object is the first element
+    const params = req.params[0];
+
+    const chainId =
+      typeof params.chainId === 'string'
+        ? (params.chainId.toLowerCase() as `0x${string}`)
+        : networkController.state.providerConfig.chainId;
+
+    // skip verification if bridge is not deployed on the specified chain.
+    // skip verification to address is not the bridge contract
+    if (
+      !Object.keys(FIRST_PARTY_CONTRACT_NAMES['MetaMask Bridge']).includes(
+        chainId,
+      ) ||
+      params.to.toLowerCase() !==
+        FIRST_PARTY_CONTRACT_NAMES['MetaMask Bridge'][chainId].toLowerCase()
+    ) {
+      return next();
```

Comment:
Assuming the intent of this middleware is to protect portfolio to bridge transactions, then won't this return allow bad extensions to redirect bridge transactions to an alternate contract?

Could we verify all portfolio transactions?